README.md
<div align="center">
  <h1>@kasa/koa-http-client</h1>
</div>

<p align="center">
  A middleware that attachs HTTP client to communicate with the context during inter-service communications in Koa.
</p>

<div align="center">
  <a href="https://circleci.com/gh/kasa-network/koa-http-client">
    <img alt="CircleCI" src="https://circleci.com/gh/kasa-network/koa-http-client.svg?style=shield" />
  </a>
  <a href="https://coveralls.io/github/kasa-network/koa-http-client">
    <img src="https://coveralls.io/repos/github/kasa-network/koa-http-client/badge.svg" alt='Coverage Status' />
  </a>
  <a href="https://www.npmjs.com/package/@kasa/koa-http-client">
    <img alt="npm version" src="https://img.shields.io/npm/v/@kasa/koa-http-client.svg" />
  </a>
  <a href="https://david-dm.org/kasa-network/koa-http-client">
    <img alt="npm" src="https://img.shields.io/david/kasa-network/koa-http-client.svg?style=flat-square" />
  </a>
  <a href="https://opensource.org/licenses/mit-license.php">
    <img alt="MIT Licence" src="https://badges.frapsoft.com/os/mit/mit.svg?v=103" />
  </a>
  <a href="https://github.com/ellerbrock/open-source-badge/">
    <img alt="Open Source Love" src="https://badges.frapsoft.com/os/v1/open-source.svg?v=103" />
  </a>
</div>

<br />

Communication between microservices must be efficient. With lots of microservices to complete a single request(transaction) successfully, each communication should have a context informations like the request id, content language, or something else. **Koa HTTP Client** will help inter-service communication with automatically attached context informations.

## Installation

```bash
# Using NPM
$ npm install --save @kasa/koa-http-client
# Using Yarn
$ yarn add @kasa/koa-http-client
```


### Dependencies

- [**Koa**](https://github.com/koajs/koa) 2.0+
- [**Node.js**](https://nodejs.org) 8.0.0+


## Usage

Use `koa-http-client` as a middleware for a [koa](https://github.com/koajs/koa) app.

```node
const Koa = require('koa');
const requestId = require('@kasa/koa-http-client');
const httpClient = require('@kasa/koa-http-client');
const app = new Koa();

app.use(requestId());
app.use(httpClient({
  retry: 3
}));
app.use(async ctx => {
  const userService = ctx.http.extend({
    baseUrl: 'http://user.company.com:8080/'
  });

  let user;
  try {
    const { body } = await userService.get('/users/john');
    user = body.user;
  } catch (err) {
    // Do something with error (HTTP status code 4xx or 5xx)
  }
  // Do something with user data
});

app.listen(3000);
```


## API

### Creating an middleware

You can create a new http client middleware by passing the relevant options to `httpClient`;

```node
const middleware = httpClient({
  forwardedHeaders: ['Accept-Language'],
  userAgent: 'user-service/0.1.0',
  requestIdHeader: 'X-Service-Request-Id',
  timeout: 20 * 1000,
  retry: 3,
});
```

### Middleware Configuration

These are the available config options for the middleware. All is optional. The middleware attachs HTTP client with the context for each request.

```node
{
  // HTTP headers to forward from Koa `ctx`
  forwardedHeaders: ['Accept-Language', 'Authorization'],

  // Client name to set `User-Agent` header
  userAgent: 'user-service/0.1.0',

  // HTTP header to get/set the request id
  requestIdHeader: 'X-Service-Request-Id',

  // Milliseconds to wait for the server to end the response before aborting the request
  timeout: 20 * 1000,

  // Count to retry on the request errors
  retry: 3,
}
```


## Credits

* [sindresorhus/got](https://github.com/sindresorhus/got): `koa-http-client` was developed based on `got`. `koa-http-client` utilizes `got.extend` feature well!


## Contributing

#### Bug Reports & Feature Requests

Please use the [issue tracker](https://github.com/kasa-network/koa-http-client/issues) to report any bugs or ask feature requests.


## License

Provided under the terms of the [MIT License](https://github.com/kasa-network/koa-http-client/blob/master/LICENSE).

Copyright © 2019, [Kasa](http://www.kasa.network).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# [1.0.0](https://github.com/kasa-network/koa-http-client/compare/v0.1.0...v1.0.0) (2019-02-22)



<a name="1.0.0"></a>
# 1.0.0 (2018-10-29)


test/middleware.test.js
'use strict';

const { EventEmitter } = require('events');
const httpClient = require('../');
const pkg = require('../package.json');


const createContext = () => ({
  req: {},
  request: {},
  res: {
    on: jest.fn()
  },
  state: {},
  get: jest.fn(),
  set: jest.fn()
});

describe('http-client', () => {
  let middleware;

  beforeEach(() => {
    middleware = httpClient();
  });

  it('should be defined as function', () => {
    expect(httpClient).toBeDefined();
    expect(httpClient).toBeFunction();
  });

  describe('middleware', () => {
    it('should be defined as koa middleware function', () => {
      expect(middleware).toBeDefined();
      expect(middleware).toBeFunction();
    });

    it('should expose a named function', () => {
      expect(middleware.name).toEqual('httpClient');
    });

    it('should attachs HTTP client into `ctx`', async () => {
      const context = createContext();

      await middleware(context, () => {
        expect(context.http).toBeDefined();
        expect(context.http).toBeFunction();
      });
    });

    it('should attachs HTTP client into `ctx` with default options', async () => {
      const context = createContext();

      await middleware(context, () => {
        const client = context.http;
        const defaultOptions = client.defaults.options;
        const defaultHeaders = defaultOptions.headers;

        expect(defaultOptions.retry.retries).toEqual(2);
        expect(defaultOptions.retry.maxRetryAfter).toEqual(30 * 1000);
        expect(defaultOptions.gotTimeout.request).toEqual(30 * 1000);
        expect(defaultOptions.json).toEqual(true);
        expect(defaultHeaders['user-agent']).toEqual(`${pkg.name}/${pkg.version}`);
      });
    });

    it('should attachs HTTP client into `ctx` with custom options', async () => {
      const middleware = httpClient({
        retry: 5,
        timeout: 10 * 1000,
        userAgent: 'helloworld'
      });
      const context = createContext();

      await middleware(context, () => {
        const client = context.http;
        const defaultOptions = client.defaults.options;
        const defaultHeaders = defaultOptions.headers;

        expect(defaultOptions.retry.retries).toEqual(5);
        expect(defaultOptions.retry.maxRetryAfter).toEqual(10 * 1000);
        expect(defaultOptions.gotTimeout.request).toEqual(10 * 1000);
        expect(defaultOptions.json).toEqual(true);
        expect(defaultHeaders['user-agent']).toEqual('helloworld');
      });
    });

    it('should delete HTTP client when the response `finish` event', async () => {
      const context = createContext();
      context.res = new EventEmitter();

      await middleware(context, () => {
        context.res.emit('finish');
        expect(context.http).toBeUndefined();
      });
    });

    it('should set the request id header if Koa `ctx` have', async () => {
      const context = createContext();
      context.id = 'my-custom-request-id';

      await middleware(context, () => {
        const client = context.http;
        const defaultHeaders = client.defaults.options.headers;

        expect(defaultHeaders['x-request-id']).toEqual(context.id);
      });
    });

    it('should delete HTTP client when the response `finish` event', async () => {
      const middleware = httpClient({
        forwardedHeaders: ['Content-Language', 'My-Custom-Header']
      });
      const context = createContext();
      const headers = new Map();
      headers.set('Content-Language', 'ko-KR');
      headers.set('My-Custom-Header', 'yeah!');
      context.get = (key) => headers.get(key);

      await middleware(context, () => {
        const client = context.http;
        const defaultHeaders = client.defaults.options.headers;

        expect(defaultHeaders['content-language']).toEqual('ko-KR');
        expect(defaultHeaders['my-custom-header']).toEqual('yeah!');
      });
    });
  });
});


lib/middleware.js
'use strict';

const got = require('got');
const debug = require('debug')('koa:http-client');
const pkg = require('../package.json');


/**
 * Return middleware that attachs a HTTP client with Koa context
 * data `ctx` for each request.
 *
 * @param {Object} [options={}] - Optional configuration.
 * @param {string[]} [options.forwardedHeaders] - HTTP headers to forward from Koa `ctx`.
 * @param {string} [options.userAgent] - Client name to set `User-Agent` header.
 * @param {string} [options.requestIdHeader] - HTTP header to get/set the request id.
 * @param {number} [options.timeout] - Milliseconds to wait for the server to end the response before aborting the request.
 * @param {number} [options.retry] - Count to retry on the request errors.
 * @param {boolean} [options.resolveBodyOnly] - Resolve body only instead of `Response` object if true.
 * @return {function} Koa middleware.
 */
module.exports = (options = {}) => {
  const {
    forwardedHeaders = ['Accept-Language'],
    userAgent = `${pkg.name}/${pkg.version}`,
    requestIdHeader = 'X-Request-Id',
    timeout = 30 * 1000,
    retry = 2,
    // TODO: `resolveBodyOnly` will be introduced in next version
    // resolveBodyOnly = true
  } = options;

  debug('Create a middleware');

  return async function httpClient(ctx, next) {
    const headers = {};

    for (const headerName of forwardedHeaders) {
      debug(`check headers['${headerName}'] to forward`);

      const header = ctx.get(headerName);
      if (header) {
        headers[headerName] = header;
        debug(`set headers['${headerName}']=${header}`);
      }
    }

    // Try to get the request id
    const reqId = ctx.id
      || ctx.state.reqId
      || ctx.req.id
      || ctx.request.id
      || ctx.reqId
      || ctx.get(requestIdHeader);

    if (reqId) {
      debug(`reqId=${reqId}`);
      headers[requestIdHeader] = reqId;
    }

    debug(`userAgent=${userAgent}`);
    headers['User-Agent'] = userAgent;

    // Attach http client to Koa `ctx` object
    ctx.http = got.extend({
      json: true,
      // responseType: 'json',
      // resolveBodyOnly,
      timeout,
      retry,
      headers
    });

    // Remove http client to mitigate accidental leaks
    ctx.res.on('finish', () => {
      debug('Remove http client for this context');
      delete ctx.http;
    });

    await next();
  };
};
